Add a check for no CredentialsRequest manifest in directory

thunderboltsid opened the pull request:
- Return an error to the user if the crederntials requests directory is empty
- Add unit tests for Nutanix create-shared-secrets subcommand

x-ref: https://issues.redhat.com/browse/CCO-108

thunderboltsid commented on pkg/cmd/provisioning/nutanix/create_shared_secrets.go:
```diff
@@ -166,7 +165,6 @@ func getCredentialsData(creds *NutanixCredentials) (*NutanixCredentials, error)
 		}
 	}
 
-	log.Printf("There are %d valid credentials", len(retCreds.Credentials))
```
This log statement causes more confusion than it helps.

thunderboltsid commented:
/retest-required

thunderboltsid commented:
/retest-required

thunderboltsid commented:
/retest-required

akhil-rane commented on pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go:
```diff
@@ -0,0 +1,199 @@
+	basicAuthCredentialsTemplate = `
+credentials:
+- type: basic_auth
+  data:
+    prismCentral:
+      username: %s
+      password: %s`
```
Should it also include auth for prism elements as defined [here](https://github.com/openshift/cloud-credential-operator/blob/master/pkg/cmd/provisioning/nutanix/create_shared_secrets.go#L48-L56)? I understand it won't affect the tests here but just for consistency and correctness.

thunderboltsid replied:
Those are optional and currently, not in use. Let me see, I can add a test that tests the contents of the output too. Then we can have one with and one without.

akhil-rane commented on pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go:
```diff
@@ -0,0 +1,199 @@
+				targetDir, err = ioutil.TempDir(os.TempDir(), testTargetDirPrefix)
+				require.NoError(t, err, "Failed to create temp directory for credentials requests")
+
+				credentialsSourceFilepath = "__does_not_exist__"
```
nit
```suggestion
credentialsSourceFilepath = "/does/not/exist"
```

akhil-rane commented on pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go:
```diff
@@ -0,0 +1,199 @@
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			credReqDir, targetDir, credentialsSource := tt.setup(t)
+			defer os.RemoveAll(credReqDir)
+			defer os.RemoveAll(targetDir)
```
do we also need to add `defer os.RemoveAll(credentialsSource)`?

thunderboltsid replied:
Good catch! Added it.

akhil-rane commented on pkg/cmd/provisioning/nutanix/create_shared_secrets_test.go:
```diff
@@ -0,0 +1,199 @@
+	}{
+		{
+			name: "No CredentialsRequest manifests in directory",
+			setup: func(t *testing.T) (credReqDir, targetDir, credentialsSourceFilepath string) {
+				credReqDir, err := ioutil.TempDir(os.TempDir(), testCredReqDirPrefix)
+				require.NoError(t, err, "Failed to create temp directory for credentials requests")
+
+				targetDir, err = ioutil.TempDir(os.TempDir(), testTargetDirPrefix)
+				require.NoError(t, err, "Failed to create temp directory for credentials requests")
+
+				credentialsDir, err := ioutil.TempDir(os.TempDir(), testCredentialsDirPrefix)
+				require.NoError(t, err, "Failed to create temp directory for credentials")
+				credentialsSourceFilepath = testBasicAuthCredentials(t, "username", "password", credentialsDir)
+				return
+			},
```
`setup` function does not return anything, I am curious how this works. Is it based on variable names?

thunderboltsid replied:
Indeed, these are named return parameters. The variables which will be returned are defined in the function signature.

akhil-rane replied:
thanks! learned something new :)

akhil-rane commented:
/lgtm

akhil-rane commented:
/assign @jeana-redhat @jianping-shu @sferich888

jeana-redhat commented:
/label docs-approved

jianping-shu commented:
@thunderboltsid @akhil-rane 
I updated the test case [OCP-48967](https://polarion.engineering.redhat.com/polarion/redirect/project/OSE/workitem?id=OCP-48967) and added step 13-15 for this PR, PTAL!
One minor issue found in step 14 is I copied the following command during the test and there was a space at the end.
./ccoctl nutanix create-shared-secrets --credentials-requests-dir=./credentials_requests --output-dir=./nutanix_secrets  
Then ccoctl will include the space in the output dir name so it became 'nutanix_secrets'$'\302\240'.
Do you think we shall enhance this?

thunderboltsid commented:
> @thunderboltsid @akhil-rane
> I updated the test case [OCP-48967](https://polarion.engineering.redhat.com/polarion/redirect/project/OSE/workitem?id=OCP-48967) and added step 13-15 for this PR, PTAL!
> One minor issue found in step 14 is I copied the following command during the test and there was a space at the end.
> ./ccoctl nutanix create-shared-secrets --credentials-requests-dir=./credentials_requests --output-dir=./nutanix_secrets
> Then ccoctl will include the space in the output dir name so it became 'nutanix_secrets'$'\302\240'.
> Do you think we shall enhance this?

Jianping and I synced offline to reproduce this. I was unable to reproduce the issue highlighted with added space and the current working theory is the bash version. @jianping-shu will try reproducing with a more recent bash version. If we manage to reproduce this, we will address this in a separate PR.

jianping-shu commented:
Talked w/ Sid in the slack, this issue has nothing to do with the PR, it will be further investigated and resolved in separate PR if need to enhance indeed.
/label qe-approved

sferich888 commented:
/label px-approved
